enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,586 @@
+
+### Risks and Mitigations
+- Risks in using google-beta provider:
+
+  Except for the additional features/functionality available for preview in the beta version,
+  google-beta provider is not different from the GA version google provider. google-beta
+  provider binary is embedded into installer as part of the installer build process and there
+  are no additional requirements made on the end user to use the beta version.
+
+  Below are some of the risks in using the google-beta provider.
+  1. google-beta is a preview version and doesn't always carry the SLAs or technical
+  support obligations.
+  2. google-beta provider always uses the beta API regardless of whether the
+  request contains any beta features.
+  3. google_compute_address API provides just the `labels` functionality in the beta version,
+  which is in beta since the initial [commit](https://github.com/hashicorp/terraform-provider-google-beta/commit/0fdc262183ed82e78824e42d05042a7b0aba7c8b)
+  in 2018, though the usual beta phase for a feature is six months.
+
+### Drawbacks
+- User-defined labels cannot be updated on an GCP resource which is not managed by an
+  operator. In this proposal, the changes proposed and developed will be part of
+  openshift-* namespace. External operators are not in scope.
+  User-defined labels can be updated on the following GCP resources.
+  - Compute Engine Instance
+  - Compute Address
+  - Compute Image
+  - Compute Forwarding Rule
+  - Storage Bucket
+  - DNS Zone
+
+- GCP Labels can be used as queryable annotations for resources, but can't be used
+  to set conditions on policies.
+
+- User-defined tags cannot be updated on a GCP resource which is not managed by an
+  operator. In this proposal, the changes proposed and developed will be part of
+  openshift-* namespace. External operators are not in scope.
+  User-defined tags can be updated on the following GCP resources.
+  - Compute Engine Instances
+  - Cloud Storage Buckets
+
+- Administrator will have to manually perform below label pertaining actions
+    1. removing the undesired labels/tags from the required resources.
+    2. update label/tags values of the required resources.
+    3. update labels/tags of the resources which are not managed by an operator.
+    4. update labels/tags of the resources for which update logic is not supported
+       by an operator.
+
+## Design Details
+
+### Open Questions
+- GCP supports 50 tags per resource and since OpenShift do not add any tags of its own,
+  users are allowed to define upto 50 tags. But the terraform and GO APIs for binding the
+  tags to the resources doesn't support batch operation and requires each tag to be added
+  individually to each resource which will increase the cluster creation time. What would
+  be the optimal number of tags a user can define?
+- GCP client libraries support two ways of providing tag details
+  1. NamespacedName: Tag value could be fed to APIs in the format
```

Comment:
This will require us to add the organization ID to the install config. This seems better than more queries to the api to get the TAG ID for each tag.

Reply:
I agree going with the NamespacedName means, we would need to make
(configured_tag_count * 2[fetch tag key + tag value]) + (configured_tag_count * resources_to_tag) and if TagValue Names is used it would be just  (configured_tag_count * resources_to_tag). And configuring an additional `Organization_ID` parameter wouldn't be required.

/cc @TrilokGeer @JoelSpeed

Reply:
I feel, it'd be better to maintain input uniformity in both install config and infra status when compared to other platforms unless there is significant delay or cost addition.

Reply:
To maintain uniformity with the solution provided for Aws, Azure tags, user will have to input Organization ID, tag key shortname and tag value shortname. The operators will use these parameters to derive the tags NamespacedName(<Organization_ID>/<tag_key_short_name>/<tag_value_short_name>)  and installer will use the tag value name fetched using the inputted params.

Reply:
`Open Questions` section updated with the solution considered for TechPreview.

Reply:
This implementation is focusing on creating tags at the organization level, but typically openshift is more concerned with the project-level and it does seem possible to create tags at the project level. 

I would not expect the installer to have permissions to create tags at the organization level.

Reply:
OpenShift will not create tags, will add the tags already existing at organization level to the resources(VM, storage bucket) created for a cluster. OpenShift will also not add tags to the project in which the resources are created.

Openshift will require permissions for viewing the tags existing at Organization level and the SA created for resources will have the permissions for adding tags to the resources.